File: packages/custos-js/src/ledger.ts
import { createHash } from "node:crypto";
import { appendFileSync, existsSync, mkdirSync, readFileSync, statSync, writeFileSync, openSync, readSync, closeSync, fsyncSync, writeSync } from "node:fs";
import { dirname, join, basename, extname } from "node:path";

import { canonicalBytes } from "./canonical.js";
import { newSpanId, newTraceId, isoNowMs } from "./ids.js";
import { KeyPair } from "./keys.js";
import {
  AttestationRecord, AttestationReason, DecisionRecord, GENESIS_PREV_HASH,
  attestationBody, recordBody,
} from "./record.js";

export class LedgerError extends Error {}

export function sha256Hex(data: Uint8Array): string {
  return createHash("sha256").update(data).digest("hex");
}

export function hashOfValue(value: unknown): string {
  return "sha256:" + sha256Hex(canonicalBytes(value));
}

function seal(record: DecisionRecord, kp: KeyPair): void {
  const body = canonicalBytes(recordBody(record));
  const hash = "sha256:" + sha256Hex(body);
  record.record_hash = hash;
  const digest = Buffer.from(hash.slice(7), "hex");
  const sig = kp.sign(digest);
  record.sig = "ed25519:" + sig.toString("base64");
}

export class Ledger {
  private _seq = 0;
  private _prev = GENESIS_PREV_HASH;

  constructor(public readonly path: string, private kp: KeyPair) {
    mkdirSync(dirname(path), { recursive: true });
    this.recoverTail();
    // pubkey sidecar
    const stem = basename(path, extname(path));
    writeFileSync(join(dirname(path), stem + ".pub"), kp.publicB64());
  }

  private recoverTail(): void {
    if (!existsSync(this.path)) return;
    const size = statSync(this.path).size;
    if (size === 0) return;
    const fd = openSync(this.path, "r");
    const readBytes = Math.min(8192, size);
    const buf = Buffer.alloc(readBytes);
    readSync(fd, buf, 0, readBytes, size - readBytes);
    closeSync(fd);
    const lines = buf.toString("utf8").split("\n").filter((l) => l.trim().length > 0);
    if (lines.length === 0) return;
    const last = JSON.parse(lines[lines.length - 1]!);
    this._seq = last.seq + 1;
    this._prev = last.record_hash;
  }

  /**
   * Append a decision record to the ledger.
   *
   * Concurrency: this method is synchronous by design — it uses openSync /
   * writeSync / closeSync and mutates in-memory chain state (seq, prev)
   * atomically within a single event-loop tick. Concurrent calls from
   * multiple event-loop tasks on the same Ledger instance therefore
   * serialize naturally (JS is single-threaded within an isolate).
   *
   * Cross-instance or cross-process concurrent writes to the same ledger
   * file are UNSAFE: seq/prev_hash state is per-instance and the hash chain
   * will diverge. Use a single Ledger per file. For higher-security
   * deployments consider re2 for policy regex evaluation as well.
   */
  append(record: DecisionRecord): DecisionRecord {
    record.seq = this._seq;
    record.prev_hash = this._prev;
    seal(record, this.kp);
    // Serialize full record (with sealed fields) canonically
    const full = { ...recordBody(record), record_hash: record.record_hash, sig: record.sig };
    const line = Buffer.concat([canonicalBytes(full), Buffer.from("\n")]);
    const fd = openSync(this.path, "a");
    try {
      writeToFd(fd, line);
      try { fsyncSync(fd); } catch { /* ignore */ }
    } finally {
      closeSync(fd);
    }
    this._seq += 1;
    this._prev = record.record_hash!;
    return record;
  }

  /**
   * Append a signed attestation record to the chain (WIRE §2.3).
   *
   * Attestations share the ledger's hash chain and signing key with
   * decisions, so a verifier reading the same file gets a single
   * timeline of "what the control was doing and whether it was alive."
   */
  appendAttestation(opts: {
    reason: AttestationReason;
    custosVersion: string;
    policyHash?: string;
    activeActors?: string[];
    uptimeMs?: number;
    traceId?: string;
  }): AttestationRecord {
    const rec: AttestationRecord = {
      v: 1,
      seq: this._seq,
      ts: isoNowMs(),
      trace_id: opts.traceId ?? newTraceId(),
      span_id: newSpanId(),
      prev_hash: this._prev,
      type: "attestation",
      attestation: {
        reason: opts.reason,
        custos_version: opts.custosVersion,
        policy_hash: opts.policyHash ?? "",
        active_actors: opts.activeActors ?? [],
        uptime_ms: opts.uptimeMs ?? 0,
      },
    };
    const body = canonicalBytes(attestationBody(rec));
    const recordHash = "sha256:" + sha256Hex(body);
    rec.record_hash = recordHash;
    const digest = Buffer.from(recordHash.slice(7), "hex");
    const sig = this.kp.sign(digest);
    rec.sig = "ed25519:" + sig.toString("base64");
    const full = { ...attestationBody(rec), record_hash: recordHash, sig: rec.sig };
    const line = Buffer.concat([canonicalBytes(full), Buffer.from("\n")]);
    const fd = openSync(this.path, "a");
    try {
      writeToFd(fd, line);
      try { fsyncSync(fd); } catch { /* ignore */ }
    } finally {
      closeSync(fd);
    }
    this._seq += 1;
    this._prev = recordHash;
    return rec;
  }

  *iterRecords(): Generator<DecisionRecord> {
    if (!existsSync(this.path)) return;
    const data = readFileSync(this.path, "utf8");
    for (const line of data.split("\n")) {
      if (!line.trim()) continue;
      yield JSON.parse(line);
    }
  }

  get seq(): number { return this._seq; }
  get head(): string { return this._prev; }
}

function writeToFd(fd: number, buf: Buffer): void {
  let offset = 0;
  while (offset < buf.length) {
    offset += writeSync(fd, buf, offset, buf.length - offset);
  }
}

File: packages/custos-py/src/custos/ledger.py
"""Append-only signed hash-chained ledger."""
from __future__ import annotations

import base64
import hashlib
import os
import threading
from pathlib import Path
from typing import Iterator, List, Optional

from custos.canonical import dumps as canonical_dumps
from custos.ids import iso_now_ms, new_span_id, new_trace_id
from custos.keys import KeyPair
from custos.record import AttestationRecord, DecisionRecord, GENESIS_PREV_HASH


class LedgerError(Exception):
    pass


def _sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def hash_of_value(value) -> str:
    """sha256:<hex> of Canonical JSON of a Python value."""
    return "sha256:" + _sha256_hex(canonical_dumps(value))


def _seal(record: DecisionRecord, keypair: KeyPair) -> None:
    body = canonical_dumps(record.to_body())
    record_hash = "sha256:" + _sha256_hex(body)
    record.record_hash = record_hash
    digest_bytes = bytes.fromhex(record_hash.split(":", 1)[1])
    sig = keypair.sign(digest_bytes)
    record.sig = "ed25519:" + base64.b64encode(sig).decode("ascii")


class Ledger:
    """Append-only JSONL ledger, one file per instance.

    Thread-safe within a process. Cross-process concurrency: open a single
    Ledger per process, or serialize through an external queue.
    """

    def __init__(self, path: str | Path, keypair: KeyPair):
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.keypair = keypair
        self._lock = threading.Lock()
        self._seq, self._prev_hash = self._recover_tail()
        # write pubkey sidecar
        pub_path = self.path.parent / (self.path.stem + ".pub")
        pub_path.write_bytes(self.keypair.public_b64().encode("ascii"))

    def _recover_tail(self) -> tuple[int, str]:
        if not self.path.exists() or self.path.stat().st_size == 0:
            return 0, GENESIS_PREV_HASH
        with self.path.open("rb") as f:
            try:
                f.seek(-4096, os.SEEK_END)
            except OSError:
                f.seek(0)
            tail = f.read().splitlines()
        for line in reversed(tail):
            if not line.strip():
                continue
            import json
            rec = json.loads(line)
            return rec["seq"] + 1, rec["record_hash"]
        return 0, GENESIS_PREV_HASH

    def append(self, record: DecisionRecord) -> DecisionRecord:
        with self._lock:
            record.seq = self._seq
            record.prev_hash = self._prev_hash
            _seal(record, self.keypair)
            line = canonical_dumps(record.to_full()) + b"\n"
            with self.path.open("ab") as f:
                f.write(line)
                f.flush()
                try:
                    os.fsync(f.fileno())
                except OSError:
                    pass
            self._seq += 1
            self._prev_hash = record.record_hash  # type: ignore[assignment]
            return record

    def append_attestation(
        self,
        reason: str,
        custos_version: str,
        policy_hash: str = "",
        active_actors: Optional[list] = None,
        uptime_ms: int = 0,
        trace_id: Optional[str] = None,
    ) -> AttestationRecord:
        """Append a signed attestation record to the chain (WIRE §2.3).

        Attestations share the ledger's hash chain and signing key with
        decisions, so a verifier reading the same file gets a single
        timeline of "what the control was doing and whether it was
        alive." No separate transport, no separate key, no separate
        integrity story.
        """
        record = AttestationRecord(
            v=1,
            seq=0,  # filled in below
            ts=iso_now_ms(),
            trace_id=trace_id or new_trace_id(),
            span_id=new_span_id(),
            prev_hash="",  # filled in below
            reason=reason,
            custos_version=custos_version,
            policy_hash=policy_hash,
            active_actors=list(active_actors or []),
            uptime_ms=uptime_ms,
        )
        with self._lock:
            record.seq = self._seq
            record.prev_hash = self._prev_hash
            body = canonical_dumps(record.to_body())
            record_hash = "sha256:" + _sha256_hex(body)
            record.record_hash = record_hash
            digest = bytes.fromhex(record_hash.split(":", 1)[1])
            sig = self.keypair.sign(digest)
            record.sig = "ed25519:" + base64.b64encode(sig).decode("ascii")
            line = canonical_dumps(record.to_full()) + b"\n"
            with self.path.open("ab") as f:
                f.write(line)
                f.flush()
                try:
                    os.fsync(f.fileno())
                except OSError:
                    pass
            self._seq += 1
            self._prev_hash = record_hash
            return record

    def iter_records(self) -> Iterator[DecisionRecord]:
        if not self.path.exists():
            return
        import json
        with self.path.open("rb") as f:
            for line in f:
                if not line.strip():
                    continue
                yield DecisionRecord.from_dict(json.loads(line))

    @property
    def seq(self) -> int:
        return self._seq

    @property
    def head(self) -> str:
        return self._prev_hash
